### PR TITLE
Add chain hash type using consts

### PR DESCRIPTION
**Pull request by tcharding**
The Lightning network defines a type called 'chain hash' that is used to uniquely represent the various Bitcoin networks as a 32 byte hash value. Chain hash is now being used by the DLC folks, as such it is useful to have it implemented in rust-bitcoin.

One method of calculating a chain hash is by hashing the genesis block for the respective network.

Add a `ChainHash` type that can be used to get the unique identifier of each of the 4 Bitcoin networks we support. Add a method that calculates the chain hash for a network using the double sha256 of the genesis block. Do so using hard coded consts and add unit tests (regression/sanity) that show these hard coded byte arrays match the hash of the data we return for the genesis block for the respective network.

The chain hash for the main Bitcoin network can be verified from LN docs (BOLT 0), add a link to this document.

Closes: #481

### CODE REVIEW

**Comment by dr-orlovsky** [comment]
Concept NACK. Nothing in LN requires chain hash to be a genesis block hash. For instance in RGB we use chain hash to be an client-side asset id.

**Comment by tcharding** [comment]
I'm not quite understanding why you ack'ed #872 and nack'ed this one, they both use genesis block as the basis for the chain hash? Am I missing something? Also, BOLT 0 specifies that the chain hash for Bitcoin mainnet must be the hash of the genesis block (https://github.com/lightningnetwork/lightning-rfc/blob/ffeece3dab1c52efdb9b53ae476539320fa44938/00-introduction.md#chain_hash).

Would you prefer if we added a call to `impl_array_newtype!` so that we get a `From<[u8; 32]` implementation for custom usage? We could also change the method name `for_network` -> `using_genesis_block`? But, as I mention above, I'm not sure _why_ we would want to do that if I have understood the BOLT correctly.

**Comment by tcharding** [comment]
Implements `Index` instead of using `impl_index_array` as is done in #805

**Comment by dr-orlovsky** [comment]
@tcharding LN BOLTs had no idea that there might be use cases for assets which operate no on the level of blockchain. However, by the nature of LN to be a multi-asset network, `chain_hash` is indeed the asset id, and not blockchain id. When you use `bitcoin mainnet` as an asset, it is defined as a `hash of genesis block`, but if you will use some other (non-bitcoin) asset based on bitcoin blockchain, you will see that it cant' be a hash of bitcoin genesis network and must be something else.

#872 allows use of custom hash values for `ChainHash` type, this PR does not.

**Comment by tcharding** [comment]
Oh, I see, I left out the call to  `impl_array_newtype` in this one. I'll push that change. Do you have a view on the `for_network` vs `using_genesis_block` naming?

**Comment by dr-orlovsky** [comment]
I think genesis block here is more relevant, since multiple assets can operate the same network (like in Liquid), while genesis block hash is definitely a for of the native asset id for the network, which is the only one

File: src/blockdata/constants.rs
Comment (dr-orlovsky) [inline src/blockdata/constants.rs:189]: Maybe we can use the approach you proposed in #870 just to avoid forgetting to re-base that PR after this one got merged?

Comment (tcharding) [reply src/blockdata/constants.rs:189]: I'm not fully happy with #870, that's why I ended up using `impl_array_newtype` so that this one is in line with the current code.

**Review by dr-orlovsky — COMMENTED** [review]
My personal preference will be to have genesis hashes taken from genesis blocks by the means of const functions, not hex values, which are hard to validate

**Comment by tcharding** [comment]
While conceptually feasible, converting `genesis_block` and `bitcoin_genesis_tx` to be const functions is non-trivial because of the usage of `script::Builder`. Do the unit tests not validate that the hex is correct?

**Review by dr-orlovsky — DISMISSED** [review]
ACK 03d5dffc19d2fce0f5e906216b8e6b6beae1c11a

@tcharding has convinced me :)

File: src/blockdata/constants.rs
Comment (tcharding) [reply src/blockdata/constants.rs:180]: Oh nice, that's heaps better than parsing the hex strings, thanks!

Comment (tcharding) [reply src/blockdata/constants.rs]: I changed the const names to `GENESIS_BLOCK_HASH_BITCOIN` also.

**Comment by tcharding** [comment]
Please re-review, force push makes considerable changes as suggested by @RCasatta (incl. change of const names).

**Review by sanket1729 — APPROVED** [review]
I left a small nit that is non-merge blocking. For future reference concerning my reviews, whenever I `nit` something, it is always non-merge blocking. 

Other maintainers are free to go ahead and the PR. The author can ignore or decide the comment based on a case of case basis. (length of PR, urgency, number of existing ACKs, available time etc).

File: src/blockdata/constants.rs
Comment (tcharding) [reply src/blockdata/constants.rs:179]: Is this even part of Bitcoin Core? From my understanding this is a LN and DLC thing.

File: src/blockdata/constants.rs
Comment (tcharding) [reply src/blockdata/constants.rs:183]: > AFAIR, there were some complaints exposing these as public constants. #568. But I like the way you have done it here.

Thanks for the link. I think this PR is ok according to that discussion, specifically:
- These consts are not public
- They are defined in `constants.rs`
- All networks are covered not just mainnet

Comment (tcharding) [reply src/blockdata/constants.rs:183]: > Unrelated to this PR:
> 
> Maybe we should use `Network::genesis_hash(&self)` API?

We did that in #872 but after review decided to optimize and use const byte arrays with unit tests to verify them.

Comment (tcharding) [reply src/blockdata/constants.rs:183]: Separate PRs because I open two alternate solutions in parallel.

**Comment by dr-orlovsky** [comment]
ACK 25f3ef40c53516aa149f91fe6e846be6b9ef328c

Not merging yet to give @Kixunil and @RCasatta opportunity to comment

**Comment by RCasatta** [comment]
In the referenced document https://github.com/lightning/bolts/blob/master/00-introduction.md we have:

> For the main chain Bitcoin blockchain, the chain_hash value MUST be (encoded in hex): 6fe28c0ab6f1b372c1a6a246ae63f74f931e8365e15a089c68d6190000000000.

while we are using reversed hex encoding like BlockHash.

eg. the following test fails
```
#[test]
    fn chain_hash_test_vector() {
        let h = ChainHash::using_genesis_block(Network::Bitcoin);
        assert_eq!("6fe28c0ab6f1b372c1a6a246ae63f74f931e8365e15a089c68d6190000000000", format!("{:x}", h));
    }
```

**Comment by tcharding** [comment]
Legend for catching that @RCasatta! I spent so much time on the first versions making sure that was correct and then dropped the ball when adding the byte arrays. Thanks man.

**Comment by tcharding** [comment]
Legend for catching that @RCasatta! I spent much time on the first versions making sure that was correct and then dropped the ball when adding the byte arrays. Thanks man.

**Comment by tcharding** [comment]
I ended up [ab]using a `sha256::Hash` in the tests to get at a `LowerHex` impl that doesn't display backwards.

**Comment by dr-orlovsky** [comment]
I start thinking that this is very LN-specific and probably should not be a part of `rust-bitcoin`, otherwise we need to do other LN-specific types here... Probably the proper place for it should be repo like https://github.com/LNP-BP/ln-types

File: src/blockdata/constants.rs
Comment (TheBlueMatt) [reply src/blockdata/constants.rs:179]: The Genesis block? I think that's a Bitcoin thing :)

File: src/blockdata/constants.rs
Comment (TheBlueMatt) [inline src/blockdata/constants.rs:185]: Can we  just say this is the Genesis hash?

**Review by TheBlueMatt — COMMENTED** [review]
LGTM. I agree with Dr Orlovsly that we don't want lightning-specific things here, but this isn't - this is a new way of accessing the Genesis block in rust-bitcoin, that works without allocating and without doing the hashing. Whether this needs a whole type to represent it, and whether that type needs to be in rust-bitcoin...I kinda don't think so, but I also don't really care/feel strongly about it.

File: src/blockdata/constants.rs
Comment (tcharding) [reply src/blockdata/constants.rs:185]: This is not restricted to being the genesis block hash (according to https://github.com/rust-bitcoin/rust-bitcoin/pull/878#issuecomment-1072315448). Re-reading that comment made me think that the word 'hash' shouldn't even be in the comment but then its in the type name so  `¯\_(ツ)_/¯`

File: src/blockdata/constants.rs
Comment (Kixunil) [inline src/blockdata/constants.rs:196]: Would `impl From` be too unreadable?

File: src/blockdata/constants.rs
Comment (Kixunil) [inline src/blockdata/constants.rs:320]: Use a macro to deduplicate? And perhaps move `assert_eq` to the common function.

File: src/blockdata/constants.rs
Comment (Kixunil) [inline src/blockdata/constants.rs:283]: A comment explaining why we stringly compare would be useful for future readers eager to clean it up.

**Review by Kixunil — DISMISSED** [review]
ACK 76166ac79872a015bbac08f468eb465044331b30

There are some nits but feel free to merge, I don't care.

File: src/blockdata/constants.rs
Comment (tcharding) [reply src/blockdata/constants.rs:179]: lol, I meant that chain hash is a LN/DLC thing and the LN spec says that the identifier for mainnet is the hash of the genesis block - hence I linked to the LN spec.

@sanket1729 would you prefer that this comment explained what the value used is but then linked to the bitcoin core source for the actual value?

File: src/blockdata/constants.rs
Comment (tcharding) [reply src/blockdata/constants.rs:185]: If we use `AssetId` then this is no longer anything to do with the LN spec's chain hash. Perhaps this comment, along with @TheBlueMatt's observation that this PR is just a new way of getting the genesis block without hashing implies that we should not add the `ChainHash` type at all and just expose a way for folk to get the genesis block without hashing so they can use it if they wish to create a `ChainHash` type themselves?

File: src/blockdata/constants.rs
Comment (tcharding) [reply src/blockdata/constants.rs:320]: Done, cheers!

File: src/blockdata/constants.rs
Comment (tcharding) [reply src/blockdata/constants.rs:196]: I think from discussion above we want to make it explicit that we are using the genesis block.

**Comment by tcharding** [comment]
Implemented @Kixunil's suggestions and rebased.

File: src/blockdata/constants.rs
Comment (sanket1729) [reply src/blockdata/constants.rs:179]: I am with without a comment. But I have a slight preference to link to the bitcoin core source. When reading a constant `GENESIS_BLOCK_HASH_BITCOIN` in file `src/blockdata/constants` in `rust-bitcoin`, I would expect it to be a bitcoin thing. It would feel odd that I need to check a lightning spec to double-check that the bitcoin genesis block is correct. 

ChainHash functions can link to lightning spec if needed.

**Review by Kixunil — APPROVED** [review]
ACK 8e29f2b49374483f50d9aba14e7190c4a3c7eea9

**Review by sanket1729 — APPROVED** [review]
ACK 8e29f2b49374483f50d9aba14e7190c4a3c7eea9. 

LGTM, I still have a very slight preference to link to bitcoin hash values instead of lightning spec. But I can live it. https://github.com/rust-bitcoin/rust-bitcoin/pull/878/commits/8e29f2b49374483f50d9aba14e7190c4a3c7eea9#r877663661

**Pull request merged**
